storage: add WITH (DISK) support to upsert sources

guswynn opened the pull request:
Part of https://github.com/MaterializeInc/materialize/issues/17067

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

- The top commit is just basic tests. @philip-stoev my plan to eventually move all upsert-related tests into this new mzcompose, and test both backends there. I wanted to keep this pr manageable. I also added this new composition to the tests ci pipeline.
- The bottom commit is the meat of the implementation; It contains:
  -  Support for `WITH (ON DISK)` in `CREATE SOURCE` (note that `ON DISK = true/false` are explicitly banned for clarity, and simplifying migrations in the future.
  - `--instance-directory` is the new optional plan required to use this feature. This is forward to `clusterd` from `envd`. Naming (and the naming of `InstanceContext`) isn't great, and I'm not attached to anything specific here)
  - The rest of the changes are mostly boilerplate piping things through. Note that error handling is lackluster, but thats a separate task in the epic. The non `ON DISK` flow is left unchanged.
  - Note that this is behind `--unsafe-mode`. Next step is to change this to an LD flag. This will happen in a followup pr, to avoid too much noise in one pr.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
   (see notion)
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

benesch commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -388,6 +388,10 @@ pub struct Args {
     /// The PostgreSQL URL for the storage stash.
     #[clap(long, env = "STORAGE_STASH_URL", value_name = "POSTGRES_URL")]
     storage_stash_url: String,
+    /// The directory of instance storage. If the process orchestrator
+    /// is being used, this can be a relative path.
+    #[clap(long, env = "INSTANCE_DIRECTORY", value_name = "PATH")]
+    instance_directory: Option<PathBuf>,
```
Agree we should try to find a better name than this! It is EC2 specific in its current incarnation, and "instance" is already a massively overloaded term. Perhaps `ephemeral_directory` or `scratch_directory` or somesuch.

benesch replied:
Actually, I'm pretty sure this needs to be process orchestrator specific. The Kubernetes orchestrator will conjure up a scratch directory upon request, by creating a volume mount of the right type.

benesch commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -646,6 +650,11 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
         OrchestratorKind::Kubernetes => {
+            if let Some(instance_directory) = &args.instance_directory {
+                if !instance_directory.is_absolute() {
+                    bail!("kubernetes orchestrator requires an absolute `--instance-directory`");
```
Validation of Kubernetes orchestrator-specific parameters should happen in the Kubernetes orchestrator constructor, not here.

benesch replied:
Hmm, actually, I think we should not do this validation at all! I can't see why it would need to be absolute for the Kubernetes orchestrator. It needs to be the _same_ for all spawned services, sure, but a relative path presumably works just fine.

benesch commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -717,6 +726,18 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
     let persist_clients = Arc::new(persist_clients);
     let orchestrator = Arc::new(TracingOrchestrator::new(orchestrator, args.tracing.clone()));
+
+    if let Some(instance_directory) = &args.instance_directory {
+        if instance_directory.exists() && !instance_directory.is_dir() {
+            bail!("`--instance-directory` must refer to a directory");
+        }
+
+        if !instance_directory.exists() {
+            std::fs::create_dir_all(instance_directory)
+                .context("failed to create `--instance-directory`")?;
+        }
+    }
```
This validation should happen inside of environmentd/lib.rs, not here, since it's not related to the CLI arguments specifically.

benesch replied:
Wait, actually, I don't think envd should be creating the scratch directory at all! It should be the individual clusters.

benesch commented on src/storage-client/src/types/sources.rs:
```diff
@@ -1725,6 +1742,21 @@ impl SourceDesc<GenericSourceConnection> {
 }
 
+impl<C> SourceDesc<C> {
+    /// lala
```
It'd be great if you could do a pass for missing docstrings before putting the PR up for review!

guswynn replied:
I usually do but I missed this one

guswynn replied:
Added a pre-push git hook locally

benesch commented on src/sql/src/plan/with_options.rs:
```diff
@@ -236,6 +236,24 @@ impl ImpliedValue for bool {
     }
 }
 
+#[derive(Default, Debug, Copy, Clone)]
+pub struct OnlyImpliedTrue(pub bool);
+
+impl TryFromValue<Value> for OnlyImpliedTrue {
+    fn try_from_value(_: Value) -> Result<Self, PlanError> {
+        sql_bail!("cannot specify value")
+    }
+    fn name() -> String {
+        "only_implied_true".to_string()
+    }
+}
+
+impl ImpliedValue for OnlyImpliedTrue {
+    fn implied_value() -> Result<Self, PlanError> {
+        Ok(OnlyImpliedTrue(true))
+    }
+}
```
I would rather not add this. This is complexity that we will likely forget to remove. It also is likely that someone will use this in a non-unsafe option in the future, at which point it becomes product debt, because we will have some special boolean options that do not support the consistent `OPTION = {true|false}` syntax. The fact that all boolean options support that syntax is a feature!

guswynn replied:
sounds good!

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -4137,6 +4154,9 @@ pub fn plan_alter_source(
                 sql_bail!("Cannot modify the IGNORE KEYS property of a SOURCE.");
             }
+            if let Some(_) = on_disk_opt {
+                sql_bail!("Cannot modify the IGNORE KEYS property of a SOURCE.");
```
Copy pasta.

benesch commented on src/storage/src/server.rs:
```diff
@@ -34,6 +35,8 @@ pub struct Config {
     pub now: NowFn,
     /// Configuration for source and sink connection.
     pub connection_context: ConnectionContext,
+    /// Other configuration for sources and sinks.
```
This is "instance" as in "storage instance" right?

```suggestion
    /// Other configuration for storage instances.
```

benesch commented on src/sql-parser/src/parser.rs:
```diff
@@ -2583,6 +2583,10 @@ impl<'a> Parser<'a> {
                 self.expect_keyword(INTERVAL)?;
                 CreateSourceOptionName::TimestampInterval
             }
+            ON => {
+                self.expect_keyword(DISK)?;
+                CreateSourceOptionName::OnDisk
+            }
```
I'm sorry for not realizing this when I made the original suggestion, but I'm wondering if we should make the option `DISK [= true]` or `SCRATCH DISK [= true]`. Just in case we need to stabilize the option. WDYT? Does that read better to y'all than "on disk"? It does to my ear, but maybe that' sjust me.

guswynn replied:
`DISK` seems good to me!

benesch commented on src/controller/src/clusters.rs:
```diff
@@ -449,7 +450,16 @@ where
-                        ]
+                        ];
+
+                        if let Some(instance_directory) = &instance_directory {
+                            args.push(format!(
+                                "--instance-directory={}",
+                                instance_directory.display()
+                            ));
+                        }
```
This doesn't seem like it will work correctly for the process orchestrator—they'll all end up sharing the same instance directory, no?

philip-stoev replied:
Within the instance directory, separate subdirectories are being created for each source -- `u1234`, `u2345` and so on. So it seems that multiple computeds can successfully coexist inside the same container

benesch reviewed (commented):
Left some comments within! I think something is funny about how this is currently plumbed. I think the orchestrators need to become aware of scratch directory-enabled services. In particular, I think `ServiceConfig` should learn a `scratch_dir: Option<usize>` parameter that specifies how big of a scratch directory to create. The process orchestrator will create this scratch directory in the argument specified by `--orchestrator-process-scratch-dir` (which will be required when the scratch dir is in use), and the Kubernetes orchestrator will create this scratch directory by mounting a volume of the right type (which will have to wait until cloud tells us what that is). Then the type of `ServiceConfig::args` should evolve from

```
&'a (dyn Fn(&BTreeMap<String, String>) -> Vec<String> + Send + Sync)
```

to

```
struct ServiceAssignment<'a> {
    ports: &'a BTreeMap<String, String>,
    scratch_dir: Option<String>,
}

&'a (dyn Fn(&ServiceAssignment) -> Vec<String> + Send + Sync)
```

and from that the controller can generate the correct `--scratch-dir` argument for the clusterd process.

philip-stoev commented:
@guswynn As a first step, I am going to manually patch a couple of CI jobs so that the new option is exercised in a few scenarios. Once the final format of the SQL syntax and the command-line option is finalized we can look at how to test this systematically. I may need an option that enables this across the board, not per-source, so that the `CREATE SOURCE` directives from the existing tests do not have to be modified.

philip-stoev commented on test/upsert/kafka-upsert-debezium-sources.td:
```diff
@@ -0,0 +1,229 @@
+! CREATE SOURCE doin_upsert
+  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-dbzupsert-${testdrive.seed}')
+  FORMAT AVRO USING SCHEMA '${schema}'
+  ENVELOPE DEBEZIUM ${arg.impl}
```
Unfortunately this approach of patching the `WITH` clause is good for a single smoketest testdrive test but does not scale. A global setting will be needed that simply enables the feature across the board for all `CREATE SOURCE` statements.

guswynn replied:
Is this something we can do in the testdrive binary itself?

guswynn replied:
How can `testdrive` discover all the create source statements and do this? Does it parse the sql?

philip-stoev commented:
Item No 1:

```
thread 'thread 'timely:work-1timely:work-2' panicked at '' panicked at 'thread 'instance directory to be there if rendering an ON DISK sourceinstance directory to be there if rendering an ON DISK source', timely:work-0', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0629d720fba03823c-1/materialize/tests/src/storage/src/render/upsert.rs' panicked at '/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0629d720fba03823c-1/materialize/tests/src/storage/src/render/upsert.rs:instance directory to be there if rendering an ON DISK source:157', :157/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0629d720fba03823c-1/materialize/tests/src/storage/src/render/upsert.rs14:
```

This apparently means that the directory does not exist?

philip-stoev commented:
ITem No 2. I see rocksdb files being created, but they never grow and are not updated, even though the sources have hundreds of thousands of records:

````
materialize@2a38dab7817d:/mzdata/rocksdb$ du --max-depth=1 -h
196K	./u45
196K	./u41
196K	./u33
196K	./u8
196K	./u4
196K	./u13
196K	./u29
196K	./u21
196K	./u17
196K	./u37
```

I see the computed process having e.g. the `LOG` files  open, but I do not think there are any writes coming to those files. @guswynn can you check the disk usage on the Mz instance that you use for development?

philip-stoev commented:
Item No 3. DROP SOURCE does not seem to clean up the respective rocksdb database.

guswynn commented:
@aljoscha this should be good to go!

To address @philip-stoev's concerns:

- Item number 1: I believe we resolved this offline and is unblocked
- Item number 2: I believe we resolved this offline. Data is not persisted until it has to be.
- Item number 3: is fixed in the latest commit (it only works for `IN CLUSTER` sources, and has some subtleties. @aljoscha we may have to improve this in the future)

Also, I added 2 LD flags: one that sets the default (this simplifies the testdrive test I added), and also one that that allows to use of `DISK` (as opposed to requiring `unsafe` mode)

For @benesch's points, all I believe are addressed. This behavior is currently now process-orchestrator only (zippy tests will need to wait)

guswynn commented:
@philip-stoev This is mostly to unblock a basic working way to use the rocksdb backing, so i can rebase https://github.com/MaterializeInc/materialize/pull/18912, so is it fine if we merge and iterate on the testing?

philip-stoev commented:
Yes, sorry for the delay, I spend like 2 days fighting a stupid thing in mzcompose.

I enabled  this by default for the entire CI in a separate branch and both the per-push CI and Nightly are green. I have also started a Release Qualification Test.

philip-stoev commented:
@guswynn can you do a git commit --amend push and force it,  just so that the CI runs again? For some reason the last run did not complete.

benesch commented on src/clusterd/src/bin/clusterd.rs:
```diff
@@ -160,6 +161,11 @@ struct Args {
     tracing: TracingCliArgs,
+
+    // === Other options. ===
+    /// The directory of instance storage.
```
```suggestion
    /// A scratch directory that can be used for ephemeral storage.
```

benesch commented on src/environmentd/src/bin/environmentd/main.rs:
```diff
@@ -365,6 +365,14 @@ pub struct Args {
         env = "ORCHESTRATOR_PROCESS_PROMETHEUS_SERVICE_DISCOVERY_DIRECTORY"
     )]
     orchestrator_process_prometheus_service_discovery_directory: Option<PathBuf>,
+    /// The directory of instance storage. If the process orchestrator
+    /// is being used, this can be a relative path.
```
```suggestion
    /// A scratch directory that orchestrated processes can use for ephemeral storage.
```

benesch commented on src/orchestrator-process/src/lib.rs:
```diff
@@ -149,6 +149,9 @@ pub struct ProcessOrchestratorConfig {
     /// that do not support Unix domain sockets (e.g., Prometheus, web
     /// browsers).
     pub tcp_proxy: Option<ProcessOrchestratorTcpProxyConfig>,
+    /// An optional top-level path that new sub-directories per-cluster will
+    /// be created in.
```
```suggestion
    /// A scratch directory that orchestrated processes can use for ephemeral storage.
```

benesch reviewed (commented):
Orchestrator changes LGTM, with some docco nits.

jkosh44 commented on src/sql-parser/tests/testdata/ddl:
```diff
@@ -1697,6 +1697,13 @@ CREATE SOURCE mz_source FROM POSTGRES CONNECTION pg (PUBLICATION = 'mz_source')
 =>
 CreateSource(CreateSourceStatement { name: UnresolvedItemName([Ident("mz_source")]), in_cluster: None, col_names: [], connection: Postgres { connection: Name(UnresolvedItemName([Ident("pg")])), options: [PgConfigOption { name: Publication, value: Some(Value(String("mz_source"))) }] }, include_metadata: [], format: None, envelope: None, if_not_exists: false, key_constraint: None, with_options: [Cr ...
 
+parse-statement
+CREATE SOURCE psychic FROM POSTGRES CONNECTION pgconn (PUBLICATION 'red') WITH (DISK);
+----
+CREATE SOURCE psychic FROM POSTGRES CONNECTION pgconn (PUBLICATION = 'red') WITH (DISK)
+=>
+CreateSource(CreateSourceStatement { name: UnresolvedItemName([Ident("psychic")]), in_cluster: None, col_names: [], connection: Postgres { connection: Name(UnresolvedItemName([Ident("pgconn")])), options: [PgConfigOption { name: Publication, value: Some(Value(String("red"))) }] }, include_metadata: [], format: None, envelope: None, if_not_exists: false, key_constraint: None, with_options: [Create ...
+
```
Can you add a test that tries to give `DISK` a value to make sure it returns an error.

guswynn replied:
It actually works with `= true` now, so ill try that!

jkosh44 commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -924,19 +940,21 @@ pub fn plan_create_source(
+        match &envelope {
+            UnplannedSourceEnvelope::Upsert { .. } => {}
+            _ => {
+                bail_unsupported!("ON DISK used with non-UPSERT/DEBEZIUM ENVELOPE");
```
I think `bail_unsupported` is fine, it provides a more structured error than `sql_bail`.

jkosh44 commented on src/sql/src/session/vars.rs:
```diff
@@ -520,6 +520,24 @@ const ENABLE_MULTI_WORKER_STORAGE_PERSIST_SINK: ServerVar<bool> = ServerVar {
+    safe: true,
+};
+
+/// Whether or not the `DISK` option in available `UPSERT` sources.
+const UPSERT_SOURCE_DISK_AVAILABLE: ServerVar<bool> = ServerVar {
```
This should be named `enable_upsert_source_disk` to match the other feature flag naming conventions.

jkosh44 commented on src/sql/src/session/vars.rs:
```diff
@@ -520,6 +520,24 @@ const ENABLE_MULTI_WORKER_STORAGE_PERSIST_SINK: ServerVar<bool> = ServerVar {
+const UPSERT_SOURCE_DISK_AVAILABLE: ServerVar<bool> = ServerVar {
+    name: UncasedStr::new("upsert_source_disk_available"),
+    value: &false,
+    description: "Whether or not the `DISK` option in available `UPSERT` sources.",
```
The description should formatted as "Feature flag indicating whether ... (Materialize)." to match the other feature flags.

jkosh44 reviewed (commented):
Couple small nits on the surfaces side.

guswynn commented:
@benesch has approved the orchestrator changes
@jkosh44 has signed off on the surfaces changes
I checked with @aljoscha, and after fixing nits, the storage changes are approved
@philip-stoev after ci passes I will merge this as there is no outstanding review or issues, as far as I can tell

jkosh44 commented on src/sql/src/session/vars.rs:
```diff
@@ -520,6 +520,25 @@ const ENABLE_MULTI_WORKER_STORAGE_PERSIST_SINK: ServerVar<bool> = ServerVar {
+    name: UncasedStr::new("enable_upsert_source_disk"),
+    value: &false,
+    description: "Feature flag indicating availability of the `DISK` \
+                  option in `UPSERT/DEBEZIUM` sources. (Materialize)",
```
Very small nit, but the period should go after `(Materialize)`.

jkosh44 commented on src/sql/src/session/vars.rs:
```diff
@@ -1753,6 +1774,16 @@ impl SystemVars {
+        *self.expect_value(&UPSERT_SOURCE_DISK_DEFAULT)
+    }
+
+    /// Returns the `upsert_source_disk_available` configuration parameter.
+    pub fn upsert_source_disk_available(&self) -> bool {
+        *self.expect_value(&ENABLE_UPSERT_SOURCE_DISK)
+    }
```
The name and comment for this function still need to change.
